docs/rst/tutorials/pkcs11_tutorial.rst

```diff
@@ -0,0 +1,135 @@
+
+.. contents:: Table of Contents
+    :depth: 2
+    :local:
+    :backlinks: none
+
+Background
+----------
+
+`ROS 2 Security <https://docs.ros.org/en/humble/Concepts/About-Security.html>`_ is implemented by means of the `DDS Security specification <https://www.omg.org/spec/DDS-SECURITY/About-DDS-SECURITY/>`_.
+The security files required by the aforementioned specification are encapsulated within the ROS 2 security enclaves.
+This approach allows to set different security policies for different nodes, processes, users, or devices by defining different enclaves with the corresponding permissions.
+However, ROS 2 currently requires that the private key used to encrypt and decrypt the enclave information is also stored within the enclave, which might be a security vulnerability.
+Hardware Security Modules (HSM) can be used in order to mitigate this vulnerability and PKCS #11 standard provides an API to interact with these devices.
+Vulcanexus Core provides the required support for HSM devices so the private key, instead of being stored in the filesystem, it could be inside the HSM device.
+
+This tutorial provides step-by-step instructions to improve ROS 2 Security of a talker/listener demo with HSM using Vulcanexus Core.
+
+Prerequisites
+-------------
+
+This tutorial does not require a physical HSM device connected to the test environment.
+Instead, a software HSM emulator is going to be installed and used to run this tutorial.
+
+Run the following commands to install the required packages:
+
+.. code-block:: bash
+
+    sudo apt update && apt install --yes --no-install-recommends \
+      softhsm2 \
+      gnutls-bin
+
+Please, also remember to source the environment in every terminal in this tutorial.
+
+.. code-block:: bash
+
+    source /opt/vulcanexus/humble/setup.bash
+
+ROS 2 Security infrastructure
+-----------------------------
+
+First of all the required security enclave to run this tutorial has to be generated using the ROS 2 CLI.
+ROS 2 enclaves are saved within a keystore generated running the following commands:
+
+.. code-block:: bash
+
+    mkdir vulcanexus_hsm_ws && cd vulcanexus_hsm_ws
+    ros2 security create_keystore vulcanexus_demo_keystore
+    ros2 security create_enclave vulcanexus_demo_keystore /talker_listener
+
+For more information about ROS 2 security infrastructure, please refer to `this tutorial <https://docs.ros.org/en/rolling/Tutorials/Advanced/Security/The-Keystore.html>`_.
+
+Generate HSM token
+------------------
+
+The private key that has been generated by ROS 2 CLI is going to be substituted by a virtual HSM using PKCS #11.
+Therefore, the ``key.pem`` auto-generated file within the enclave is going to be replaced by a ``key.p11`` file with the PKCS #11 URI.
+As already explained, this tutorial emulates the HSM using ``softHSM``.
+Generate the HSM token running:
+
+.. code-block:: bash
+
+    softhsm2-util --init-token --free --label vulcanexus_demo --pin VulcanexusDemo --so-pin vulcanexus_demo
```

Comment:
I run the tutorial within a docker so maybe privileged permissions are required. I could not run the tutorial in my laptop because there are no binaries for Vulcanexus Humble in Focal. I can run the tutorial in Vulcanexus Galactic and see if I have any trouble. I do not remember anything related to a specific group...